### ISSUE TITLE
help

### ISSUE THREAD

**Issue by gg680** [Henry-Lee-real/MaeFuse#8]
Hello, the paper mentions the cfm and mfm modules, two stages of training, freezing parameters and so on, but in your code is directly through the fusion loss, and does not jointly optimize the entire network, is it a problem with my understanding of the code?

**Comment by Henry-Lee-real** [comment]
Our code doesn’t yet include a full end-to-end pipeline (each time I’ve been tweaking the previous version instead of rebuilding from scratch). In the **train** section I explained how you could structure the code if you want to run the whole process. Give that a try first; if it’s still unclear, I’ll carve out some time to write a fully detailed script for you.

**Comment by gg680** [comment]
thanks ，I'm trying

**Comment by Henry-Lee-real** [comment]
If you still haven’t figured it out after a week, feel free to let me know again.

**Comment by gg680** [comment]
I'm still a little confused.    Here is my own modification of the two-stage training code        

 if epoch < args.epochs // 2:
            # if epoch > args.epochs // 2:
                result, vi_latent, ir_latent, templefusion = connect(imgs_vi, imgs_ir, return_latent=True, freeze_encoder=True)
                latent_mean = (vi_latent + ir_latent) / 2

                # print(templefusion.shape)
                # print(latent_mean.shape)
                loss_total = torch.mean((templefusion - latent_mean) ** 2)
                loss = loss_total
                loss_out_fusion += loss.item()

                scaler.scale(loss).backward()
                scaler.step(optimizer)
                scaler.update()
                count += 1

          
            else:
            
                result, vi_latent, ir_latent, templefusion = connect(imgs_vi, imgs_ir, return_latent=True,
                                                                     freeze_encoder=True)
                recover = trans_img(result, std, mean)  # output_imgs
                imgs_vi = trans_img(imgs_vi, std, mean) #vi_imgs
                imgs_ir = trans_img(imgs_ir, std, mean) #ir_imgs

                loss_total, loss_grad, loss_laplacian = Loss(recover, imgs_vi, imgs_ir)

                loss = torch.mean(loss_total)

                loss_out_fusion += torch.mean(loss_total).item()
                loss_out_grad += torch.mean(loss_grad).item()
                loss_out_laplacian += torch.mean(loss_laplacian).item()

                scaler.scale(loss).backward()  

> 

                scaler.step(optimizer)  
                scaler.update()  
                count += 1     

Can you walk me through the details of my modifications? I'm having some problems with the second phase of training. The losses won't go down.

**Comment by Henry-Lee-real** [comment]
Did you freeze the CFM module during the second stage of training, or could you show me a graph of the loss decreasing?

**Comment by gg680** [comment]
This is the loss function for my two-stage training process, but there is a problem with my image. The horizontal axis is trained once every 175 rounds, and there is an error in my loss fluctuation, but I don't know where the problem lies.

<img width="802" height="657" alt="Image" src="https://github.com/user-attachments/assets/5e8fc62f-b256-4806-aaa2-0249ff85c3c4" />

**Comment by Henry-Lee-real** [comment]
Please translate the following passage. Additionally, does one epoch comprise both stages executed sequentially (Stage I followed by Stage II)?

**Comment by gg680** [comment]
No, the image depicts two-stage training. Prior to this, I had already completed the first stage of training and frozen the parameters from that stage.

**Comment by Henry-Lee-real** [comment]
Could you take a look at the fused images after restoration? If they can be fused normally, I think it should be fine. After the final training stage finishes, you can enable (unfreeze) the decoder and fusion layers and train them together for about 10 epochs.
